javalib/src/main/scala/java/util/AbstractList.scala

```diff
@@ -1,3 +1,284 @@
+          val oIter = o.listIterator
+          this.forall(oIter.hasNext && _ === oIter.next) && !oIter.hasNext
+        case _ => false
+      }
+    }
+  }
+
+  override def hashCode(): Int = {
+    this.foldLeft(1) {
+      (prev, elem) => 31 * prev + (if (elem == null) 0 else elem.hashCode)
+    }
+  }
+
+  protected def removeRange(fromIndex: Int, toIndex: Int): Unit = {
+    @tailrec
+    def removeNext(index: Int, iter: ListIterator[E]): Unit = {
+      if (iter.hasNext && index < toIndex) {
+        println(iter.next())
+        iter.remove()
+        removeNext(index + 1, iter)
+      }
+    }
+    removeNext(fromIndex, listIterator(fromIndex))
+  }
+
+  protected[this] def checkIndexInBounds(index: Int): Unit = {
+    if (index < 0 || index >= size)
+      throw new IndexOutOfBoundsException(index.toString)
+  }
+
+  protected[this] def checkIndexOnBounds(index: Int): Unit = {
+    if (index < 0 || index > size)
+      throw new IndexOutOfBoundsException(index.toString)
+  }
+
+  private abstract class AbstractListView(fromIndex: Int,
+        private var toIndex: Int) extends AbstractList[E] {
+
+    override def add(e: E): Boolean = {
+      add(size, e)
+      toIndex += 1
+      true
+    }
+
+    override def add(index: Int, e: E): Unit = {
+      checkIndexOnBounds(index)
+      self.add(fromIndex + index, e)
+    }
+
+    override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+      checkIndexOnBounds(index)
+      self.addAll(fromIndex + index, c)
+      val elementsAdded = c.size
+      toIndex += elementsAdded
+      elementsAdded != 0
+    }
+
+    override def addAll(c: Collection[_ <: E]): Boolean =
+      addAll(size, c)
+
+    override def clear(): Unit = {
+      self.removeRange(fromIndex, toIndex)
+      toIndex = fromIndex
+    }
+
+    def get(index: Int): E = {
+      checkIndexInBounds(index)
+      self.get(fromIndex + index)
+    }
+
+    override def remove(index: Int): E = {
+      checkIndexInBounds(index)
+      val elem = self.remove(fromIndex + index)
+      toIndex -= 1
+      elem
+    }
+
+    override def set(index: Int, e: E): E = {
+      checkIndexInBounds(index)
+      self.set(fromIndex + index, e)
+    }
+
+    def size(): Int =
+      toIndex - fromIndex
+
+  }
+
+  /*
+   * RandomAccessListIterator implementation assumes that the has an efficient
+   * .get(index) implementation.
+   */
+  private class RandomAccessListIterator(private var i: Int, start: Int,
+        private var end: Int) extends ListIterator[E] {
+
+    private var last = -1
+
+    def hasNext(): Boolean =
+      i < end
+
+    def next(): E = {
+      last = i
+      i += 1
+      get(last)
+    }
+
+    def hasPrevious(): Boolean =
+      start < i
+
+    def previous(): E = {
+      last = i
+      i -= 1
+      get(i)
+    }
+
+    def nextIndex(): Int = i
+
+    def previousIndex(): Int = i - 1
+
+    def remove(): Unit = {
+      if (i == -1 || i == last)
+        throw new IllegalStateException
```

Comment:
Then of course `i == last` here is incorrect. Instead, I would reset `last = -1` a few lines below instead of `last = i`.